src/Raven.Client/Documents/InMemoryDocumentSessionOperations.cs

```diff
@@ -814,13 +820,13 @@ private void PrepareForEntitiesPuts(SaveChangesData result)
             {
                 UpdateMetadataModifications(entity.Value);
                 var document = EntityToBlittable.ConvertEntityToBlittable(entity.Key, entity.Value);
-                if ((entity.Value.IgnoreChanges) || (!EntityChanged(document, entity.Value, null))) continue;
+                if (entity.Value.IgnoreChanges || (!EntityChanged(document, entity.Value, null)))
+                    continue;
 
-                if (!_documentStore.BeforeStoreEmpty)
-                {
-                    _documentStore.OnBeforeStore(this, entity.Value.Id, entity.Key, GetMetadataFor(entity.Key));
+                var beforeStoreEventArgs = new BeforeStoreEventArgs(this, entity.Value.Id, entity.Key);
+                OnBeforeStore?.Invoke(this, beforeStoreEventArgs);
+                if ((OnBeforeStore != null) && EntityChanged(document, entity.Value, null))
```

Comment:
Why aren't you checking the beforeStoreEventArgs.EntitySerializionRequired here?